Add `JsonInclude.Value` convenience constants

runeflobakk opened the pull request:
See discussion here: https://github.com/FasterXML/jackson-databind/discussions/5294

This pull-request adds some constants to `JsonInclude.Value` equivalent to the enum values of `JsonInclude.Include`.

This e.g enables the following for configuring a JsonMapper:
```java
var jsonMapper = JsonMapper.builder()
    .defaultPropertyInclusion(JsonInclude.Value.ALL_NON_NULL)
    .build();
```


I noticed a kind of established pattern in `JsonInclude.Value` to declare `static final` constants as `protected`, and use a static _method_ to access it. If this pattern is also applicable for these constants, just let me know, and I will edit the pull-request accordingly!

runeflobakk committed:
Add JsonInclude.Value constants

Equivalent to the enum values of JsonInclude.Include.
See also discussion here:
https://github.com/FasterXML/jackson-databind/discussions/5294

cowtowncoder commented on src/main/java/com/fasterxml/jackson/annotation/JsonInclude.java:
```diff
@@ -262,6 +262,49 @@ public static class Value
+        /**
+         * Value that indicates that property is to be always included,
+         * independent of value of the property.
+         */
```
Add `@since 2.21`

cowtowncoder committed:
Merge branch '2.x' into 2.x

cowtowncoder commented on src/main/java/com/fasterxml/jackson/annotation/JsonInclude.java:
```diff
@@ -268,6 +268,49 @@ public static class Value
 
+        /**
+         * Value that indicates that property is to be always included,
+         * independent of value of the property.
```
Should probably expand a bit to indicate this applies to both "value" and "content" cases (can refer to class Javadoc) -- similarly for all new constants.

runeflobakk replied:
Is the general paragraph I added ok?

cowtowncoder replied:
Ah, no this is bit different than what I meant actually: not between annotating Class vs annotating property, but for case of `java.util.Map` (and `Optional`) where:

1. `value` means inclusion/exclusion of `Map` itself/reference (exclude if `null`, empty etc)
2. `content` means inclusion of key/value entries of `Map`: if Map entry with `null` value, empty value etc are included (and for contained value in case of `Optional`).

So, f.ex:

```
Map<String,Integer> m = Map.of("key", null);
```

will be serialized as

```
{ "key", null }
```

regardless of `@JsonInclude(value = XXX)`, if "content" is left as default (`Include.ALWAYS`).
But if we have `@JsonInclude(content = Include.NON_NULL)`
it will be serialized as

```
{ }
```

And it is completely omitted if `@JsonInclude(value = Include.NON_EMPTY, content = Include.NON_NULL)` -- by first filtering out `"key": null` (null-valued entry), and then Map itself (since no entries included, it's empty).

Ok that's not the best possible explanation... but I hope it gives the idea.

(note, tho: "content" does not apply to any other types; not for `Collection`s, Arrays or POJOs -- for those, "value" only matters)

runeflobakk replied:
Hmmm, I admit I struggle to understand what you mean 😅 , because what you describe _is_ already my understanding too and what I attempted to describe, and the phrases I used are deliberately taken more or less verbatim from the javadoc of `JsonInclude`. But maybe I have misunderstood that documentation.

> 1. value means inclusion/exclusion of Map itself/reference (exclude if null, empty etc)

This was what I referred to as "including a value (...) [on Java object level](https://github.com/FasterXML/jackson-annotations/blob/2.x/src/main/java/com/fasterxml/jackson/annotation/JsonInclude.java#L17)". I interpret "object level" as describing members such as fields and accessors, and their values (e.g. a String or a Map itself).

> 2. content means inclusion of key/value entries of Map: if Map entry with null value, empty value etc are included (and for contained value in case of Optional).

This I referred to as "[when contained](https://github.com/FasterXML/jackson-annotations/blob/2.x/src/main/java/com/fasterxml/jackson/annotation/JsonInclude.java#L24) in an [object reference](https://github.com/FasterXML/jackson-annotations/blob/2.x/src/main/java/com/fasterxml/jackson/annotation/JsonInclude.java#L19-L22)".

Since this distinction is already documented on `JsonInclude`, I wanted to make sure I did not duplicate this documentation, but tried to express it in (very) condenced terms, with a reference to the more detailed description of "value" vs. "content" wrt. inclusion in serialization.

But please, can you have a go at what you would prefer the text to be? :)

cowtowncoder replied:
Yeah sorry, this is area where I struggle as well. Thank you for explaining what you meant.
I only explained logic to make sure we agree -- which we do (great!).
I think one challenge is that existing description is not great to begin with; I tried straighen it a bit but not sure with what effect.

runeflobakk replied:
In this context, the word "value" is overloaded quite a bit. Both used to refer to values being candidates for serialization, and the rule itself being named `Value`. Would it be helpful to change the first sentence in the javadoc  to e.g:
```diff
< Value that indicates that property is to be always included, independent of value of the property.
> Constant that indicates that property is to be always included, independent of value of the property.
```
To be a bit more precise when referring to the actual instance of `Value`, vs. values being included in serialized form. (And a "value" is thought of being in the shape of either a _value_ or _contained_, as specified by a `Value`.)

Maybe `JsonInclude.Value` would be better off named as `JsonInclude.Rule`, just to decrease the amount of overloading. But that's of course just food for thought (maybe for 3.x? I haven't looked into how these concepts are represented there).

cowtowncoder replied:
Yes, that rephrasing would be a good start.

And yes, it is unfortunate "value" is overloaded (and to a degree, has to be, it being Annotation default property -- originally `@JsonInclude` only had "value" property, no "content, allowing `@JsonInclude(NON_NULL)` compact notation).
Plus convention I chose wrt Annotation companion classes for `jackson-annotations`.

Renaming of `JsonInclude.Value` is not something I think should be done, given how it's convention across Jackson annotations (could not be done for 2.x for compatibility, but IMHO should not be done for 3.0 even if technically allowed).

runeflobakk replied:
[426a065](https://github.com/FasterXML/jackson-annotations/pull/314/commits/426a065707334aac87ec423063a4e14ffcee2000)

cowtowncoder commented:
I think it's ok to just expose these particular constants without accessor.
Added couple of minor comments, but looks good overall.

One process thing: I'd need CLA:

https://github.com/FasterXML/jackson/blob/main/contributor-agreement.pdf

(unless you have sent one in the past, if so, LMK).

The usual way is to print, fill & sign, scan/photo, email to `cla` at fasterxml dot com. This only needs to be done once for any and all contributions.
Once I get it I can merge the PR (pending review/approval) etc.

runeflobakk committed:
Add missing version spec in javadoc: 2.21

runeflobakk committed:
Add javadoc details about value inclusion

The constants specify the same JsonInclude.Include for inclusion of
values both on Java object level, as well when contained in an object
reference (e.g. an AtomicReference).

cowtowncoder commented:
NOTE: CLA received.

runeflobakk committed:
Rephrase to use "Constant" instead of "Value"

When referring to the instances of JsonInclude.Value, to avoid too much
overloading with "values" referring to values being serialized to JSON.

See also discussion:
https://github.com/FasterXML/jackson-annotations/pull/314#discussion_r2328334071

cowtowncoder committed:
Add release notes

cowtowncoder commented on src/main/java/com/fasterxml/jackson/annotation/JsonInclude.java:
```diff
@@ -268,6 +268,84 @@ public static class Value
+         * This will specify the same setting for including a value both
+         * on <b>Java object level</b> as well as when <b>contained</b>
+         * in an object reference (see {@link JsonInclude} for further
+         * details on this distinction).
```
Actually makes sense & looks good. Thank you for doing this.

cowtowncoder reviewed (approved):
LGTM
